browser/playlist/playlist_data_source.cc

```diff
@@ -39,8 +55,69 @@ scoped_refptr<base::RefCountedMemory> ReadFileToString(
       contents.length());
 }
 
+content::URLDataSource::RangeDataResult ReadFileRange(
+    const base::FilePath& file_path,
+    net::HttpByteRange range) {
+  CHECK_CURRENTLY_NOT_ON_UI_THREAD();
+
+  base::File file(file_path,
+                  base::File::Flags::FLAG_OPEN | base::File::FLAG_READ);
+  if (!file.IsValid()) {
+    return {};
+  }
+
+  // Note that HTTP range's first and last position are inclusive.
+  int64_t first_byte_position =
+      range.HasFirstBytePosition() ? range.first_byte_position() : 0;
+  int64_t last_byte_position = range.HasLastBytePosition()
+                                   ? range.last_byte_position()
+                                   : std::numeric_limits<uint32_t>::max();
+  int64_t read_size = std::min(kMediaChunkSizeInByte,
+                               last_byte_position - first_byte_position + 1);
+
+  auto buffer = base::HeapArray<uint8_t>::Uninit(read_size);
+  auto read_result = file.Read(first_byte_position, buffer);
+  if (!read_result.has_value()) {
+    return {};
+  }
+  read_size = read_result.value();
+
+  content::URLDataSource::RangeDataResult result;
+  result.buffer = base::MakeRefCounted<base::RefCountedBytes>(buffer);
+  result.file_size = file.GetLength();
+  result.range = net::HttpByteRange::Bounded(
+      first_byte_position, first_byte_position + read_size - 1);
+  auto mime_type = playlist::mime_util::GetMimeTypeForFileExtension(
+                       file_path.FinalExtension())
+                       .value_or("video/mp4");
+  result.mime_type = mime_type;
+  return result;
+}
+
 }  // namespace
 
+PlaylistDataSource::DataRequest::DataRequest(const GURL& url) {
+  const auto full_path = content::URLDataSource::URLToRequestPath(url);
+  const auto paths = base::SplitStringPiece(
+      full_path, "/", base::TRIM_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+  CHECK_EQ(paths.size(), 2u);
+
+  id = paths.at(0);
+  const auto& type_string = paths.at(1);
+  if (type_string == "thumbnail") {
+    type = kThumbnail;
+  } else if (type_string == "media") {
+    type = kMedia;
+  } else if (type_string == "favicon") {
+    type = kFavicon;
+  } else {
+    NOTREACHED() << "type is not in {thumbnail,media,favicon}/ : "
```

Comment:
Not sure if this was intentional
```suggestion
    NOTREACHED() << "type is not in {thumbnail,media,favicon}: "
```